From d6543ff78c57d93152b2831631dba4a94c32e58b Mon Sep 17 00:00:00 2001
From: Alex Belits <alex.belits@cavium.com>
Date: Thu, 18 Oct 2018 10:46:11 +0300
Subject: [PATCH 0316/1051] octeontx: Add context / ASID locking in GPIO
 interrupt handlers.

Signed-off-by: Alex Belits <Alex.Belits@cavium.com>
Signed-off-by: Yury Norov <ynorov@marvell.com>
Signed-off-by: Yury Norov <ynorov@caviumnetworks.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
index 29d946d25cec..35a058280da9 100644
--- a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
+++ b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
@@ -16,6 +16,7 @@
 #include <linux/cdev.h>
 #include <linux/uaccess.h>
 #include <linux/device.h>
+#include <linux/mmu_context.h>
 
 #include "octeontx.h"
 #include "octeontx_mbox.h"
@@ -1092,6 +1093,7 @@ static inline int __install_el3_inthandler(unsigned long gpio_num,
 
 	spin_lock_irqsave(&el3_inthandler_lock, flags);
 	if (!gpio_installed[gpio_num]) {
+		lock_context(current->group_leader->mm, gpio_num);
 		arm_smccc_smc(THUNDERX_INSTALL_GPIO_INT, gpio_num,
 			      sp, cpu, ttbr0, 0, 0, 0, &res);
 		if (res.a0 == 0) {
@@ -1101,6 +1103,8 @@ static inline int __install_el3_inthandler(unsigned long gpio_num,
 			gpio_installed_tasks[gpio_num]
 				= current->group_leader;
 			retval = 0;
+		} else {
+			unlock_context_by_index(gpio_num);
 		}
 	}
 	spin_unlock_irqrestore(&el3_inthandler_lock, flags);
@@ -1120,6 +1124,7 @@ static inline int __remove_el3_inthandler(unsigned long gpio_num)
 		gpio_installed[gpio_num] = 0;
 		gpio_installed_threads[gpio_num] = NULL;
 		gpio_installed_tasks[gpio_num] = NULL;
+		unlock_context_by_index(gpio_num);
 		retval = 0;
 	} else {
 		retval = -1;
-- 
2.17.1

